src/Orleans.Core/Networking/ConnectionManager.cs

```diff
@@ -383,42 +366,55 @@ private IDisposable BeginConnectionScope(Connection connection)
             return null;
         }
 
-        private struct ConnectionEntry
+        private class ConnectionEntry
         {
-            public readonly DateTime? LastFailure;
-            public readonly ImmutableArray<Connection> Connections;
-            private readonly int[] lockObj;
+            public Task PendingConnection { get; set; }
+            public DateTime? LastFailure { get; set; }
+            public ImmutableArray<Connection> Connections { get; set; }
```

Comment:
Connections can be accessed outside of a lock, so we use an immutable collection here.